File: 1-discrete-math/09-lab-linear-data-structures/src/A.6/stack1.py
class vector:
	def __init__(self):
		self.l = 8
		self.cnt = 0
		self.a = [0] * self.l
		
	def push(self, new):
		if self.cnt == self.l:
			b = [0] * (2 * self.l)
			b[:self.l] = self.a
			self.a = b
			self.l *= 2
		self.a[self.cnt] = new
		self.cnt += 1
	
	def pop(self):
		if self.cnt < self.l // 4 and self.cnt > 4:
			self.l //= 2
			self.a = self.a[:self.l] 
		self.cnt -= 1
		return self.a[self.cnt]

	def top(self):
		return self.a[self.cnt - 1]

next = open('stack1.in').readline
out = open('stack1.out', 'w').write
n = int(next())
a = vector()

for i in xrange(n):
	s = next()
	if s[0] == '-':
		out(a.pop())
	else:
		a.push(s[2:])

File: 1-discrete-math/09-lab-linear-data-structures/src/A.4/stack1.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.StringTokenizer;

public class stack1 {
	static class Stack {
		final int INIT_LENGTH = 8;
		String[] a;
		int len;
		int cnt;
		
		public Stack() {
			a = new String[INIT_LENGTH];
			len = INIT_LENGTH;
		}
		
		void push(String s) {
			if (len == cnt) {
				String[] b = new String[len * 2];
				for (int i = 0; i < cnt; i++) {
					b[i] = a[i];
				}
				len *= 2;
				a = b;
			}
			a[cnt++] = s;
		}
		
		String pop() {
			if (cnt < len / 4 && cnt > INIT_LENGTH) {
				String[] b = new String[len / 2];
				for (int i = 0; i < cnt; i++) {
					b[i] = a[i];
				}
				len /= 2;
				a = b;
			}
			return a[--cnt];
		}
	}
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new FileReader("stack1.in"));
		int n = Integer.parseInt(br.readLine());
		Stack s = new Stack();
		PrintWriter out = new PrintWriter("stack1.out");
		for (int i = 0; i < n; i++) {
			StringTokenizer st = new StringTokenizer(br.readLine());
			if (st.nextToken().equals("+")) {
				s.push(st.nextToken());
			} else {
				out.println(s.pop());
			}
		}
		out.close();
	}
}
